telemetry/prometheus/convert.go

```diff
@@ -0,0 +1,158 @@
+package prometheus
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-live-sdk/telemetry"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+var _ telemetry.Converter = (*Converter)(nil)
+
+// Converter converts Prometheus metrics to Grafana frames.
+type Converter struct {
+	format expfmt.Format
+}
+
+// NewConverter creates Converter.
+func NewConverter() *Converter {
+	return &Converter{
+		format: expfmt.FmtText,
+	}
+}
+
+func (c Converter) metricsToFrameWrappers(mfs []*dto.MetricFamily) ([]telemetry.FrameWrapper, error) {
+	var frameWrappers []telemetry.FrameWrapper
+	for _, mf := range mfs {
+		metricType := mf.GetType()
+		metricName := mf.GetName()
+		fields := make([]*data.Field, 3)
+		fields[0] = data.NewField("time", nil, []time.Time{})
+		fields[1] = data.NewField("label", nil, []string{})
+		switch metricType {
+		case dto.MetricType_COUNTER, dto.MetricType_SUMMARY, dto.MetricType_GAUGE:
+			fields[2] = data.NewField("value", nil, []float64{})
+		case dto.MetricType_HISTOGRAM:
+			fields[2] = data.NewField("value", nil, []uint64{})
+		}
+		for _, m := range mf.GetMetric() {
+			// TODO: handle m.GetTimestampMs()
+			switch metricType {
+			case dto.MetricType_COUNTER:
+				counter := m.GetCounter()
+				fields[0].Append(time.Now())
+				fields[1].Append(labelsString(m.Label))
+				fields[2].Append(counter.GetValue())
+			case dto.MetricType_GAUGE:
+				gauge := m.GetGauge()
+				fields[0].Append(time.Now())
+				fields[1].Append(labelsString(m.Label))
+				fields[2].Append(gauge.GetValue())
+			case dto.MetricType_SUMMARY:
+				summary := m.GetSummary()
+				// TODO: sum and count.
```

Comment:
The only idea that came to me is make sum and count as separate frames, but it seems so wasteful